### PR TITLE
async-1: Stub ChunkLoader

### PR DESCRIPTION
**Pull request by pwinston**
# Description
* Image changes and stub `ChunkLoader`
* See more async PRs: [The Napari Rendering GitHub Project](https://github.com/users/pwinston/projects/1)

# Details
* Adds `napari.components.chunk` module with stub `ChunkLoader`, `ChunkKey` and `
ChunkRequest`
* `ChunkLoader` is a stub that always loads **synchronously**.
* Add `Layer.loaded` property and `loaded` event
    * By default layers are permanently in the loaded state
* Adds `Image._load_images()` and `Image.chunk_loaded()` methods.
* Small `ImageView` change.
* New `ImageSlice.load_chunk()` and `ImageSlice.chunk_loaded()` methods.
* It loads **synchronously** 100% of the time.

### CODE REVIEW

**Comment by pwinston** [comment]
I added a few tests just to get the balling rolling a bit. Much more more are needed. Not sure how to turn code coverage on for this branch?

**Comment by pwinston** [comment]
I think this is ready for review @jni.

File: napari/components/chunk/_tests/test_chunk.py
Comment (jni) [inline napari/components/chunk/_tests/test_chunk.py]: We use absolute imports in tests.

```suggestion
from napari.layers.image import Image
from napari.components.chunk import ChunkKey, chunk_loader
```

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py]:
```suggestion
        The combined key, all the identifiers together.
```

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py]: These should be `Optional[int]`s, since any of them can be None:

```python
In [1]: x = slice(4, 8)

In [2]: x.step is None
Out[2]: True
```

File: napari/layers/image/_image_slice.py
Comment (jni) [inline napari/layers/image/_image_slice.py:49]: Reading this, I am wondering what the difference is between an ImageSlice and an ImageView...

File: napari/layers/image/_image_slice.py
Comment (jni) [inline napari/layers/image/_image_slice.py]:
```suggestion
        # This will return a satisfied request if ChunkLoader is doing
        # synchronous loading or the chunk was in the cache. If it returns
```

File: napari/layers/image/_image_slice.py
Comment (jni) [reply napari/layers/image/_image_slice.py:49]: ... which upon further investigation reminds me of how I objected to having too many different objects back in #1343 but my objections were ignored, and now I am paying the price. =P

File: napari/layers/image/_image_slice.py
Comment (jni) [inline napari/layers/image/_image_slice.py]: See #1355. Now is an excellent time to implement it.

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py:17]: I'm a little hesitant to have this class. After watching the video for the whole architecture, I thought, ok, maybe my dislike of new objects is a little too dogmatic, and this isn't so bad. But now upon reviewing this, I am finding myself confused by ImageSlice vs ImageView and ImageConverter. To me, the takeaway is that we should strive to use simple Python types as much as possible.

Now, in this case, the conversion of a tuple to a key is not *quite* trivial, so I'm willing to get some pushback on this, but I'd certainly like to moot the idea of using a single function for this, `chunk_to_key()`. It seems pretty similar ultimately, but I am newly entranced by the idea of losing an extra class definition.

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py:106]: This is interesting — is it correct? I'm wondering whether the indices in key itself need transposing...?

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py]: I think this can be removed in this PR. I know it seems pointless to remove to only add later, and I'm sorry, but I think it's part of the idea of keeping this PR small and simple. That will make it clearer in the later PR why it needs adding.

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py:73]: I'm *really* confused about why this is a dictionary?

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py:575]: What is a thumbnail source?

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py:573]: Given my proposal above to drop ChunkKey, this would change to:

```python
self._load_images(chunk_to_key(self, image_indices), image, thumbnail_source)
```

I'm biased but I like it. ;)

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py:559]: Ah, so is the idea that ImageSlice is the slice, and ImageView has a possible transposition? Is that the only difference?

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py:633]: Why is this here? Seems unlikely to ever be needed...?

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py]: Should this be maybe renamed `_on_chunk_loaded`? It looks like this is only ever called by `self`?

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py]: As mentioned above, I'd prefer to drop this from this PR.

File: napari/components/chunk/_request.py
Comment (jni) [reply napari/components/chunk/_request.py:73]: It looks like this contains just two keys, `image` and `thumbnail`. Is there a vision for more? It's funny because I'm always advocating for simple Python objects, and now here it really feels like these should be attributes rather than a dictionary.

File: napari/layers/image/image.py
Comment (jni) [inline napari/layers/image/image.py]: Ok so reading this I'm wondering:

- can we replace `chunk_loader` with a simple `concurrent.futures.ThreadPoolExecutor()`
- instead of `self._slice.load_chunk(request)` we do `future = executor.submit(self._slice.load)`
- since we are doing sync for now, we immediately do `self._slice.set_data(future.result())`

Would this work, or am I missing some steps?

I'm not necessarily trying to get rid of the ChunkLoader, but rather, I want to defer it until it is actually needed. My suspicion is that we might end up with a simpler design if we simplify maximally in this PR. In particular, ChunkRequest is quite lightweight and I'm wondering whether we can do without it (same as ChunkKey) and have it be replaced by Futures having (key, Array) inputs and outputs.

**Review by jni — COMMENTED** [review]
Thanks @pwinston! This is still a major PR but was reasonable to review. I'm wondering whether it can be minimised further, specifically:

- ChunkKey -> chunkid_to_key()
- ChunkRequest -> concurrent.futures.Future
- ChunkLoader -> Executor

There's more minor comments throughout but that's the gist.

**Comment by jni** [comment]
I forgot to say: thanks for the hard work ripping all this out for review! I know it's extra work but I think everyone on the team will end up understanding the finished product much more clearly!

File: napari/layers/image/_image_slice.py
Comment (pwinston) [reply napari/layers/image/_image_slice.py:49]: From the comments an `ImageSlice` is "The slice of the image that we are currently viewing" while an `ImageView` is "A raw image and a viewable version of it.". Those are literally the first words in the doc strings of the two classes.

In napari's `Image` layer there are images and there are thumbnails of those images. So when you are viewing slice of an image, the `ImageSlice` is the image slice that you are viewing, and it has two attributes: `self.image` and `self.thumbnail`, both of which are of the type `ImageView`.

Each `ImageView` meanwhile has two properties `self.view` and `self.raw` both of which are `ArrayLike`. The raw one is the raw image and the view one is the viewable able, which was converted from raw to viewable using the `ImageConverter` (now called the `Callable[[ArrayLike], ArrayLike]` converter). 

So an `ImageSlice` contains two `ImageViews`, and each `ImageView` contains two images. This exact behavior and data existed in `Image` long before I made any changes, but it was all in the base class `Image`. Does that make sense?

File: napari/layers/image/_image_slice.py
Comment (pwinston) [reply napari/layers/image/_image_slice.py]: I removed `ImageConverter` in this PR, and commented at #1355.

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py:17]: I think `ChunkKey` gains more functionality in the following PR's. This PR was extracted from #1510 and the stripped down. Although actually it does not gain much. Still I'd like to keep it as-is while we focus on splitting and fixing issues that are preventing merging.

If we change it now all the following PR's change. I don't totally disagree with this suggestion, but it seems more style and could be easily changed some other time once all the parts are in, if it in fact we still see it as a problem.

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py:106]: It is required to transpose the images. This was always done. We can move the transposing to someone where, but it has to be done somewhere. Are you saying you want to move it to somewhere else? If you don't think it's required try commenting it out!

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py]: I removed it but yes it will be immediately added back in the in the next PR.

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py:73]: The `ChunkLoader` knows nothing about images. The `Image` layer currently asks the `ChunkLoader` to load 1 or 2 chunks. However other layers might load different chunks with different names, and other types of images might load more chunks with different names in one request.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py:575]: A thumbnail source is the image that's used to generate the thumbnail. This was in `Image` class long before I started, so nothing was changed.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py:559]: No, explained in comment above.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py:633]: I hit this in practice so I added this check. Without this check you get a very cryptic error.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: Maybe, but I wasn't user if the `_on_` prefix implied something to do with events? Also this is called externally by `QtChunkReceiver` in #1510 so it can't start with a leading underscore. But it could be `on_chunk_loaded` if `on` does not mean event related?

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: I removed it.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: See #1510 for how the `ChunkLoader` grows. Your suggestion at the bottom of #1510 was "add the changes to Image/ImageSlice, calling out to a dummy `ChunkLoader` object/module that simply loads things synchronously" so that's what I've done here. I say we stick with the plan, since no new information has come to light since the plan was made.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py:573]: I think that's a valid way to do it, but that it's a question but it's a matter of style. This PR was split out from #1510. We could re-write this in a different style but then we'd have to redo every other PR that's part of #1510. I'd like to focus on splitting up #1510 and fixing the issues that are preventing #1510 from being merged first.

I think a style pass to remove OO would not be crazy at all. Especially if the Napari Style Guide discourages OO. But I think combining that style pass with simply splitting up the PRs would be mixing the two efforts, and would complicated the splitting.

File: napari/layers/image/_image_slice.py
Comment (pwinston) [reply napari/layers/image/_image_slice.py]: Fixed!

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py]: Fixed, thanks.

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py]: Fixed, thanks.

File: napari/components/chunk/_tests/test_chunk.py
Comment (pwinston) [reply napari/components/chunk/_tests/test_chunk.py]: Fixed, is there a reason tests do it differently? Is this in the Napari Style Guide?

**Comment by pwinston** [comment]
Addressed all @jni comments, fixed or replied to them. Thanks!

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py]: @pwinston True! But the point of the plan is to be able to more clearly catch such simplifying steps in review, not to just end up with the exact same code at the end...

Right now, it looks to me like we might not need a lot of the machinery that you are setting up hook points for. Once the hooks are there, it will be tricky in the next PR to see what was necessary versus what was there for pre-review historical reasons.

On the other hand, if we go the simple route now, in the next PR you can go, "well, now we need a place to simultaneously hold the threadpoolexecutor and the queue, so that's the ChunkLoader". And I'll be like, well, fair enough. Or, maybe another simplification will occur to me in that context...!

So, I'd like to discuss whether this simplification makes sense, rather than whether it matches the exact plan we laid out.

At any rate, I'll refer to the "future version" of the chunkloader in #1510, good point, but that'll have to wait until tomorrow.

File: napari/layers/image/_image_slice.py
Comment (jni) [reply napari/layers/image/_image_slice.py:49]: Yeah, makes sense, thanks for the explainer. I'll need to think about things more — I feel like the class names are not descriptive enough. I'm old school and I think code should be readable without jumping back and forth between class definitions. =) That's a separate issue — I'll raise it.

File: napari/components/chunk/_request.py
Comment (jni) [reply napari/components/chunk/_request.py:106]: OHHHHHH Sorry — I was confusing `key` with ChunkKey, as opposed to image and thumbnail. 👍

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: To put it another way, the reason it looks like the `ChunkLoader` is not doing much is your suggestion at the bottom of #1510 was explicitly "create a preliminary version of the `ChunkLoader` that's a stub, that does not do very much". So it's a stub (that could be replaced) only because the goal of the PR was to make a stub.

File: napari/components/chunk/_request.py
Comment (jni) [reply napari/components/chunk/_request.py]: As mentioned elsewhere, this PR should have a minimum of extraneous stuff. As it stands it's very close! Thanks for removing it.

File: napari/components/chunk/_request.py
Comment (jni) [reply napari/components/chunk/_request.py:73]: Could we roll this into the ChunkKey? (Whether or not we turn that into a function.) The extra level of indirection confused me repeatedly while reviewing this PR...

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py:575]: @pwinston this method is new I believe? Either way it should have a proper NumPy docstring that describes what the arguments are.

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py:573]: > Especially if the Napari Style Guide discourages OO.

Oooh I like this idea! The Juan Style Guide certainly does. Let's codify it. =P

@pwinston how are you envisioning the coming steps? If you are cherry-picking earlier commits, I think merge conflicts are going to start arising just from all the other suggested changes. If not, then I think the removal of ChunkKey at least would not be that additionally onerous to maintain, since it's a simple text replacement.

In general, though, I know a simple copy-paste for the breakdown is appealing, but the point of the breakdown into smaller PRs is to see along the way how we can simplify, and *only add complexity when and where we need it*. So although I acknowledge it might be more effort, I think we'll end up in a better place.

One idea: can you open the second PR against this PR (ie on your repo), so that I can judge for myself what I'm ok with keeping? =)

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py:633]: Does this not mean that it should get checked in the image init? I agree that things should be caught early, but this is actually really not that early! 😂

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py]: It generally *is* what's done on events, but it's also kind of what makes sense in English... This function is called when a chunk has been loaded... Right? Maybe `when_chunk_loaded`? `after_chunk_loaded`? `finalize_loaded_chunk`?

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py]: Yup yup, I'm just saying, the stub might be even more minimal, to the point that it could be removed. Sometimes these things aren't clear at the outset!

**Comment by jni** [comment]
@pwinston I think the best approach to resolve some of the discussions above is to open async-2 against async-1 in your own repo and ping me here. =) 

Also, please let reviewers mark conversations as resolved. Otherwise it is very hard for me to see your responses — and these are useful for other reviewers as well.

File: napari/components/chunk/_tests/test_chunk.py
Comment (jni) [reply napari/components/chunk/_tests/test_chunk.py]: This will test the installed version, while in modules/packages you want to have a *consistent* version and that means relative imports. You could argue that you want to test a consistent version also, but generally tests are not necessarily in packages and relative imports are forbidden in non-packages.

**Comment by pwinston** [comment]
Sorry, I thought the author was supposed to mark them resolved. So if the reviewer asks for a trivial change, and you verbatim make their change, you still let the reviewer mark it resolved? That seems like pushing work on to them? Is that a universal GitHub thing or a Napari thing? 

I think "should we remove this class" when it's not against any written guideline is best handled with a top-level issue for several reasons:

1. This issue is more visible.
2. These are exactly the types of issues that people need to see and learn from, since they are judgment calls.
3. The PR from resolving _that small issue_ will be incredibly bite-sized and targeted, a better more focused PR than even this PR (which is in fact big).
4. The change is visible in the record as a PR, as a focused change, so people can see that, it's not buried inside this PR.
3. Anyone can resolve that issue and by definition it's absolutely perfect for first timer, since it's small and focused and into existing functionality that works.

The idea that "this is our last opportunity to make changes to the code" I think is a bad idea to foster about reviews. And it's bad to promote the idea that only the author can make the changes, even if those changes are totally mechanical and easy, like a name change. I think the best time to address style questions is after some use of the new code. Like "I was adding async support to Surface, while the `ChunkLoader` work I think it this part was awkward, here's a PR to improve it."

I can create create `async-2` next week and maybe ping you next friday, so you can continue your review of `async-1` the following monday, which is a US Holiday, so the first we can merge `async-1` is September 2nd, which is a few days after my current contract is over. 

Or we can merge it now and you file a nice cogent top-level issue, or 2 or 3, with your suggestions, I start on async-2 this weekend and maybe get async-3 done next week as well. My preference is make the concrete changes now, and I made all the concrete changes you requested, and file _issues_ on the fuzzier _issues_, and move on to the next 4 PR's which are backed up.

Napari should, and does to my knowledge, have systems in place so that well-written issues will be addressed _just as reliably_ as buried review comments, and top-level issues are a better platform for visibility, learning, and enticing contributors into helping on juicy easy changes.

I would call focusing on the most important problems in the PR and filing issues on style/judgement concerns async reviewing. Which in the end will get you the same result much faster than synchronous reviewing, where people are blocked for weeks at a time waiting. I think async reviewing is particularly the right choice for an opt-in feature, soon to be called an `experimental` feature.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py:633]: This problem is new with async loading. With async loading we don't have the image data until it has been loaded. At that time it's possible to discover the user set the "rgb" flag incorrectly. If we don't explicitly check here, it leads to a nasty confusing error. So I added the check you see here.

That is I went through this experience:
1) I had a test case that produced a nasty confusing error.
2) I added this check and re-ran the test case.
3) I saw the error was now reasonable.
4) I fixed the test case.

I felt that was a good approach.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: I changed it to `on_chunk_loaded` which is definitely better than `chunk_loaded`. But maybe it could be confusing, I'm not sure.

File: napari/layers/image/image.py
Comment (pwinston) [reply napari/layers/image/image.py]: The stub could be removed. In fact that was my very first suggestion to you on the call. I said Juan, we don't even need the `ChunkLoader`, we can just modify the `Image` class to not call into anything, to do the sync load right there.

And you said no, Phil, we do need the `ChunkLoader` in the initial PR but make it a stub, it's just a stub in the first version that does sync loads. So in the initial PR that I split out of #1510 I created a stub `ChunkLoader` as we discussed as you asked for. I think it's fine, because it's all about to be stomped with the next PR, if we can finish thing PR.

Doing a 180 degree reversal now, just so that we can do a 2nd 180 degree reverse in three days, would be a colossal waste of time. That's the issue. As you can see in #1510 the ChunkLoader is there in every step. It's like 500 lines of code.

My strong advisement with splitting up #1510 is:
1) In each split PR find every serious problem and fix them **before merging**.
2) Style and design issues wait until the parts are merged, because they are parts of whole.
3) Everything is merged. It's opt-in, it's experimental, no one is using it.
4) File issues on every remaining concern, no matter how minor, and they will be addressed.

I feel you completely disagree with the above steps? Yet I don't understand what steps you are attempting to follow. It seems you simply don't believe that the code can ever be changed once it's in master? Is that the case? Because if that's true then it does change things.

**Comment by sofroniewn** [comment]
> So if the reviewer asks for a trivial change, and you verbatim make their change, you still let the reviewer mark it resolved? That seems like pushing work on to them? Is that a universal GitHub thing or a Napari thing?

I resolve trivial, uncontroversial changes where I have very clearly done what was requested. If I have made a comment not wanting to do something, or done an alternative I leave those for the reviewer to resolve if they want. Sometimes we merge with those comments left open if, we decided to do something else, but want to leave that discussion as very easily discoverable.

> Or we can merge it now and you file a nice cogent top-level issue, or 2 or 3, with your suggestions, I start on async-2 this weekend and maybe get async-3 done next week as well. My preference is make the concrete changes now, and I made all the concrete changes you requested, and file issues on the fuzzier issues, and move on to the next 4 PR's which are backed up.

This seems like the right approach to me. I will do a quick pass through the code right now. @jni will be up soon too and can hopefully give one last green light.

Removing a class inside the chunk loader part of the code can easily be done later.

**Review by sofroniewn — APPROVED** [review]
I have just read through, focusing on the image layer code - which is the most connected to the rest of the code base, and so where changes will have the most impact on other PRs and efforts.

I also wonder long-term about `ChunkKey` as an object, and how and where it gets used, but I think its absolutely fine to get this in now and then continue to iterate and improve in subsequent PRs.

I'd like @jni to be able to have one final look at this in the morning when he wakes up, but then I'd like to merge so as to not leave @pwinston blocked.

**Comment by jni** [comment]
Just to clarify: I'm not blocking anything right now, async-2 can and should be open right now against this branch. Then if we immediately merge this branch it is a simple matter of opening it against master rather than this branch.

**Comment by pwinston** [comment]
> I'm not blocking anything right now.

Above you said you are blocking the merge of `async-1` until I make `async-2`. That's very definition of blocking. If you could explain what you meant by "I'm not blocking anything" and what you mean by "anything" in that sentence? It makes me wonder how much else is getting lost in translation.

The problem with your approach is then `async-1` is in limbo, you are asserting it's in limbo until I create `async-2`, and I have to create `async-2` not knowing whether `async-1` will require more changes. Then when you see `async-2` you want to go back to `async-1` and possibly make changes, which will then result in changes in `async-2`.  To me this totally subverts all of our motivations for splitting up the PR, and I really could not dream up a worse, slower or more unproductive way to review a big change.

The best approach in my eyes is merge the split PR's in series, in order. Review each one, fix all major issues, merge it. Review the next one, fix all major issues, merge it. Then with the experimental opt-in code in master I'm happy to entertain 100 issues and iterate on it endlessly.

It's like we have a 7 car pile up here. This is not a normal situation. We are trying to untangle the pile up. We came up with a plan of splitting up the PR. Let's follow through with that plan. Fix major issues while merging everything. Then come up for air and figure out what's next, whether re-work and re-design or fix the gross visuals problem or what.

I'm happy to keep iterating on async code, even just what's in #1510, as long as CZI wants to fund me. But at this point the most efficient way to use those funds is get this stuff merged then iterate. I think there's some huge disconnect here and I'm not sure what they right way is to get to the bottom of it.

**Comment by jni** [comment]
@pwinston yeah, sorry about this, I totally understand how you would be reluctant to make async-2 until this is final. I was under the impression that you have async-2 in your back pocket, so you could grab it and use it to convince me that most of my concerns are unfounded.

Also and btw, sorry about the somewhat curt message — I was on my phone and merely wanted to get async-2 surfaced, rather than wait for a weekend cycle.

> Review each one, fix all major issues, merge it. 

For me, the object proliferation *is* a major issue. However, I can see that the whole sequence is going to be too hard to re-engineer. At least having the PRs separated will make review easier, and it will also make it easier to revisit these design choices after the whole train is merged. It's like a written record of your video. So 👍 on that for sure.

> > So if the reviewer asks for a trivial change, and you verbatim make their change, you still let the reviewer mark it resolved? That seems like pushing work on to them? Is that a universal GitHub thing or a Napari thing?
> 
> I resolve trivial, uncontroversial changes where I have very clearly done what was requested. If I have made a comment not wanting to do something, or done an alternative I leave those for the reviewer to resolve if they want. Sometimes we merge with those comments left open if, we decided to do something else, but want to leave that discussion as very easily discoverable.

👆 like that. We should probably codify this in our contributing guide. The reason it's not is historical — the bulk of it (based on scikit-image's) was written before the Resolve button existed!

> Napari should, and does to my knowledge, have systems in place so that well-written issues will be addressed just as reliably as buried review comments

Well, this is called technical debt as you know. Like real debt, a little is good and necessary, a lot means you can go bankrupt. You might be right that the balance is a little too conservative right now, I don't know. But if you look at other packages in the ecosystem, the intuition develops that the *likely* fate of an issue is that it won't get resolved, especially if it is not blocking critical functionality.

I *also* think that the PR is indeed the best time to fix code issues, because that is when the code is fresh in your mind. I'm pretty confident speculating that everyone reading this has had the experience of revisiting old code they wrote and thinking, now what the hell was I doing here...? 

Anyway, having said all of the above, I'm happy to move forward — let's take on that debt! I'll write up the ~~loan contracts~~ issues for the post-async world. 😊

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py:633]: > I felt that was a good approach.

It is! I am just wondering whether there is an even better one lurking. For example, having the first request go through a slightly different code path than subsequent ones. Does that sound reasonable to you? It just feels super weird to me to have this check every time we get a request back. Is it worth me raising an issue? (Thoughts @sofroniewn?)

File: napari/layers/image/image.py
Comment (jni) [reply napari/layers/image/image.py]: Just to clarify: yes, I said something on the call based on partial understanding. Thanks to the actions we agreed on in the call, my understanding has improved (I think), which means I was now advocating a slightly different approach. There is no paradox.

> It seems you simply don't believe that the code can ever be changed once it's in master? Is that the case? Because if that's true then it does change things.

Well, code does have pretty massive inertia. It's not that it can never be changed, it's that it *might* never be changed. Resources are finite. Anyway, we'll see how we go! 🤞

**Comment by pwinston** [comment]
I think if people feel the PR review process is the only opportunity to add value to the codebase, and if people feel that any issues they file are going to be ignored, no matter how valid the problem, then those are both pretty big problems. I don't have any quick answers. But it'd be interesting to compare notes and ideas.

**Comment by sofroniewn** [comment]
Glad we got this merged @pwinston @jni, thanks for all the effort to get this in! I think it's onwards and upwards from here, with lots of lessons learnt and things to continue to think over and improve!!!

File: napari/layers/image/image.py
Comment (sofroniewn) [reply napari/layers/image/image.py:633]: >  It just feels super weird to me to have this check every time we get a request back. Is it worth me raising an issue? (Thoughts @sofroniewn?)

I agree, something feels funny here, but it could represent something deeper than rgb/ async that we want to consider even more now. We do have it so that you can change the dimensionality of the image data by modifying the data property, similarly the multiscale/ rgb too, but I'm not sure if we should allow that - i.e. once you create a layer should the dimensionality/ rgb/ multiscale be fixed, this could dramatically simplify lots of things, while being a reasonable restriction in functionality. In the absence of this restriction, I think the checks that @pwinston has here might be critical and we might need more for "dimensionality changes" too.

A dedicated issue to continue discussion seems the right approach @jni

Comment (d-v-b) [reply napari/layers/image/image.py:633]: What was the nasty confusing error? Maybe that could be made less confusing.

Comment (pwinston) [reply napari/layers/image/image.py:633]: I don't remember the error, it was a few months ago. But it was not ours. Just something from numpy about brocasting dimensions or mis-aligned shapes. Something that gave zero indication it had anything to do the with the fact you lied about whether your image was rgb or not. Such as you said the image was not rgb, but it was, or vice verse. Without async that's found out immediately, with async it's not. I agree a top-level issue might be the best way to sort it out. Perhaps someone else can take an interest and come up with a better solution.

Comment (pwinston) [reply napari/layers/image/image.py:633]: I suspect there are a lot more issues like this that we won't find until we get heavy testing, like people using it all day for routine work. Something we are aiming towards. I feel like my usage is VERY gingerly, I'm just focusing on the happy path, seeing the async logic work, and not on all the many edge cases.

Comment (jni) [reply napari/layers/image/image.py:633]: Raised in #1576. I won't resolve this so that the reference to this discussion loads correctly.

**Pull request merged**
